File: 21560 - Sai Suraj/week 5/684_Redundant Connection.py
class DSU:
    def __init__(self):
        self.par = [i for i in range(1001)]
        self.rank = [0] * 1001

    def find(self, x):
        if self.par[x] != x:
            self.par[x] = self.find(self.par[x])
        return self.par[x]

    def union(self, x, y):
        xr, yr = self.find(x), self.find(y)
        if xr == yr:
            return False
        elif self.rank[xr] < self.rank[yr]:
            self.par[xr] = yr
        elif self.rank[xr] > self.rank[yr]:
            self.par[yr] = xr
        else:
            self.par[yr] = xr
            self.rank[xr] += 1
        return True
    
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        dsu = DSU()
        for edge in edges:
            if not dsu.union(*edge):
                return edge

File: 21551-naveen-siva-kumar/week5/684-redundant-connection/684-redundant-connection.java
class Solution {
    public int[] parent;
    public int cycle;
    
    
    public int[] findRedundantConnection(int[][] edges) {
        //using union find in java
        int n = edges.length;
        int universe[] = new int[n+1];
        for(int i=1;i<n+1;i++){
            universe[i]=i;
        }
        makeUniverse(universe);
        
        for(int i=0;i<n;i++){
            union(edges[i][0],edges[i][1],i);
        }
        return edges[cycle];
        
    }
    
    public void makeUniverse(int[] universe){
        this.parent = new int[universe.length];
        //this cycle = new ArrayList<int>();
        for(int i=1;i<universe.length;i++){
            parent[i]=i;
        }
    }
    
    public void union(int node1,int node2,int index){
        
        int x = find(node1);
        int y=find(node2);
        
        if(parent[x]==y){
            //cycle.add(index);
            this.cycle=index;
            return;
        }
        else{
            parent[x]=y;
            return;
        }
    }
    public int find(int node){
        if (parent[node] == node) {
            return node;
        }
        return find(parent[node]);
    }
}